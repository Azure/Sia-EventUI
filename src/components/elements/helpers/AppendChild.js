import React from 'react'

const AppendChild = (component, childToAppend) => React.cloneElement(
  component,
  {
    children: component.props.children
      ? Array.isArray(component.props.children)
        ? component.props.children.concat(childToAppend)
        : [component.props.children, childToAppend]
      : childToAppend
  }
)

export default AppendChild