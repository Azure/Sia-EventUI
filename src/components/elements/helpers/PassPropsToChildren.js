import React from 'react'

const PassPropsToChildren = (children, props) => React.Children.map(
  children,
  child => child.type && child.type !== 'div'
    ? React.cloneElement(child, props)
    : React.cloneElement(
      child,
      {
        children: PassPropsToChildren(
          child.props.children,
          props
        )
      }
    )
)

export default PassPropsToChildren
