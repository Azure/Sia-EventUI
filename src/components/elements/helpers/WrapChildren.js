import React from 'react'

export const WrapChildren = (component, wrapper) => React.cloneElement(
  component,
  {
    children: React.Children.map(
      component.props.children,
      (child, index) => wrapper(child, index)
    )
  }
)

export default WrapChildren
