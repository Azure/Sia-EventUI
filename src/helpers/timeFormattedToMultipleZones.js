import { DateTime } from 'luxon'
import _ from 'underscore'

const zones = [
  { shortname: 'PT', ianaZone: 'America/Los_Angeles'},  // PST https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
  { shortname: 'IST', ianaZone: 'Asia/Kolkata'},         // India's IANA zone https://en.wikipedia.org/wiki/Time_in_India
  { shortname: 'GMT', ianaZone: 'Etc/GMT'}
]

const format = Object.assign(DateTime.DATE_SHORT, DateTime.TIME_24_WITH_SECONDS)

export const timeFormattedToMultipleZones = (time, timezones = zones) => {
  let dateFormat = { year: "numeric", month: "numeric", day: "numeric" }
  let convertTimeToZone = (timepoint) => {
    timepoint.timeInZone = time.setZone(timepoint.ianaZone)
    return timepoint
  }

  let timeInMultipleZones = _.chain(timezones)
      .map(convertTimeToZone)
      .groupBy((timepoint) => timepoint.timeInZone.toLocaleString(dateFormat))
      .map((dateGroup) => {
        return _.map(dateGroup, (timepoint, index) => {
          debugger
          return index === 0 ?
                  timepoint.timeInZone.toLocaleString(format) + ' ' + timepoint.shortname :
                  timepoint.timeInZone.toLocaleString(DateTime.TIME_24_WITH_SECONDS) + ' ' + timepoint.shortname
        }).join(', ')
      })
      .compact()
      .value()
      .join('; ')
  // let timeInMultipleZones = _.chain(timezones).map(convertTimeToZone).value().join('; ')
debugger
  return timeInMultipleZones
}

export default timeFormattedToMultipleZones

// () => {
//   timeInZone.toLocaleString(format)
//
// return formatted_time + ' ' + timezone.shortname
// }

// dateGroup) => {
//   return _.map(dateGroup, (datetime, date) => {
//     debugger
//     return _.first(dateGroup[date]) === datetime ?
//               datetime.toLocaleString(format) :
//               datetime.toLocaleString(DateTime.TIME_24_WITH_SECONDS)
//   }).join(' lejoiner ')
// }
