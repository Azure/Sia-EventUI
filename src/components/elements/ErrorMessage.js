import React from 'react'
import { DateTime } from 'luxon'
import ErrorIcon from 'material-ui/svg-icons/alert/error'
import { Card, CardHeader } from 'material-ui/Card'
import { RetryButton } from 'components/elements/Buttons'

import timeFormattedToMultipleZones from 'helpers/timeFormattedToMultipleZones'

export const ErrorMessage = ({
  message,
  actionForRetry,
  time = null,
  backgroundColor = null
}) => {
  const errorMessageTime = time && time instanceof DateTime ? time.toLocal().toFormat(DateTime.TIME_WITH_SECONDS) : null

  return <Card
    className='incident-card'
    style={{ backgroundColor }}
  >
    <ErrorIcon />
    <CardHeader
      title={message}
      subtitle={timeFormattedToMultipleZones(errorMessageTime)}
    />
    { actionForRetry ? <RetryButton actionForRetry={actionForRetry} /> : null }
  </Card>
}

export default ErrorMessage
