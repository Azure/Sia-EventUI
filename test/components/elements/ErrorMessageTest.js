'use strict'
import React from 'react'
import { expect } from 'chai'
import { DateTime } from 'luxon'
import { shallow } from 'enzyme'
import 'test/helpers/configureEnzyme'
import toJson from 'enzyme-to-json'
import { describeSnapshot } from 'test/helpers/describeSnapshot'

import ErrorMessage from 'components/elements/ErrorMessage'

describe('ErrorMessage', function () {
  context('when inputs are valid', function () {
    describe('when given only a message', function () {
      const testObject = shallow(<ErrorMessage message={'TestMessage'} />)

      describeSnapshot(function () {
        it('should match snapshot', function () {
          expect(toJson(testObject)).to.matchSnapshot()
        })
      })
    })

    describe('when given a message and provided an action for retry', function () {
      const testObject = shallow(<ErrorMessage message={'TestMessage'} actionForRetry={'TestAction'} />)

      describeSnapshot(function () {
        it('should match snapshot', function () {
          expect(toJson(testObject)).to.matchSnapshot()
        })
      })
    })

    describe('when given a time', function () {
      const testTime = DateTime.fromObject({
        year: 2018,
        month: 4,
        day: 25,
        hour: 15,
        minute: 39,
        second: 22,
        zone: 'America/Los_Angeles'
      })

      const testObject = shallow(<ErrorMessage time={testTime} />)

      describeSnapshot(function () {
        it('should match snapshot', function () {
          expect(toJson(testObject)).to.matchSnapshot()
        })
      })
    })

    describe('when given a background color', function () {
      const backgroundColor = 'Purple'
      const testObject = shallow(<ErrorMessage backgroundColor={backgroundColor} />)
      describeSnapshot(function () {
        it('should match snapshot', function () {
          expect(toJson(testObject)).to.matchSnapshot()
        })
      })
    })
  })
})
