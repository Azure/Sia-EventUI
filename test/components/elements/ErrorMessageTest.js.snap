// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorMessage when inputs are valid when given a background color snapshot should match snapshot 1`] = `
<Card
  className="incident-card"
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
  style={
    Object {
      "backgroundColor": "Purple",
    }
  }
>
  <AlertError />
  <CardHeader
    avatar={null}
    subtitle={
      <Connect(TimeDisplayComponent)
        time={null}
      />
    }
  />
</Card>
`;

exports[`ErrorMessage when inputs are valid when given a message and provided an action for retry snapshot should match snapshot 1`] = `
<Card
  className="incident-card"
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
  style={
    Object {
      "backgroundColor": null,
    }
  }
>
  <AlertError />
  <CardHeader
    avatar={null}
    subtitle={
      <Connect(TimeDisplayComponent)
        time={null}
      />
    }
    title="TestMessage"
  />
  <Connect(DisplayRetryButton)
    actionForRetry="TestAction"
  />
</Card>
`;

exports[`ErrorMessage when inputs are valid when given a time snapshot should match snapshot 1`] = `
<Card
  className="incident-card"
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
  style={
    Object {
      "backgroundColor": null,
    }
  }
>
  <AlertError />
  <CardHeader
    avatar={null}
    subtitle={
      <Connect(TimeDisplayComponent)
        time={"2018-04-25T15:39:22.000-07:00"}
      />
    }
  />
</Card>
`;

exports[`ErrorMessage when inputs are valid when given only a message snapshot should match snapshot 1`] = `
<Card
  className="incident-card"
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
  style={
    Object {
      "backgroundColor": null,
    }
  }
>
  <AlertError />
  <CardHeader
    avatar={null}
    subtitle={
      <Connect(TimeDisplayComponent)
        time={null}
      />
    }
    title="TestMessage"
  />
</Card>
`;
